### PR TITLE
SONARJAVA-1377 Fine tuning

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java [hunk 1]
```diff
@@ -305,19 +319,55 @@ public boolean noReturnDescription() {
...
+        // exceptions used in javadoc is using simple name when method declaration use fully qualified name
+        descriptions = thrownExceptions.get(toSimpleName(exceptionName));
+      }
+      if (descriptions == null) {
+        // exceptions used in javadoc is using fully qualified name when method declaration use simple name
+        descriptions = thrownExceptions.entrySet().stream()
+          .filter(e -> toSimpleName(e.getKey()).equals(exceptionName))
+          .map(Map.Entry::getValue)
+          .flatMap(List::stream)
+          .collect(Collectors.toList());
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java]: MAJOR Unknown method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

File: java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java [hunk 2]
```diff
@@ -305,19 +319,55 @@ public boolean noReturnDescription() {
...
     public List<String> undocumentedThrownExceptions(Tree tree) {
-      return getUndocumentedElements(getExceptions(tree), thrownExceptions);
+      List<String> exceptionNames = getExceptions(tree);
+      if (exceptionNames.size() == 1 && "Exception".equals(exceptionNames.get(0)) && !thrownExceptions.isEmpty()) {
+        // check for described exceptions when only "Exception" is used is declared as being thrown
+        return thrownExceptions.entrySet().stream()
+          .filter(e -> isEmptyDescription(e.getValue()))
+          .map(Map.Entry::getKey)
+          .map(Javadoc::toSimpleName)
+          .collect(Collectors.toList());
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java]: MAJOR Unknown method [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=java-debugging-rules%3AUnknownMethodInvocation)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java [hunk 3]
```diff
@@ -367,23 +396,23 @@ private static boolean isEmptyDescription(@Nullable String part) {
...
+          return ((IdentifierTree) typeTree).name();
         case MEMBER_SELECT:
-          return ((MemberSelectExpressionTree) typeTree).identifier();
+          return ExpressionsHelper.concatenate((MemberSelectExpressionTree) typeTree);
         default:
-          throw new IllegalStateException(
-            "Should never happen. Exceptions can not be specified other than with an identifier or a fully qualified name, and can not be parametrized.");
+          // should not happen, as exceptions can only be defined using identifier or fully qualified name,
+          // no parameterized type can be used
+          return null;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java]: So why actually returning something ? I would really prefer to stick to throwing an explicit exception. AFAI understand you only return so you can test it in an existing test : create a dedicated test for the case and keep the explicit failure.

File: java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java [hunk 4]
```diff
@@ -445,24 +447,51 @@ private static String extractMainDescription(List<String> lines) {
...
+      Map<String, List<String>> results = new HashMap<>();
+      for (int i = 0; i < lines.size(); i++) {
+        String line = lines.get(i);
+        Matcher matcher = pattern.matcher(line);
+        if (matcher.matches()) {
+          String elementName = matcher.group("name");
+          if (!results.containsKey(elementName)) {
+            results.put(elementName, new ArrayList<>());
+          }
+          List<String> currentDescriptions = results.get(elementName);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java]: use results.computeIfAbsent(elementName, ArrayList::new)

**Pull request merged**